sentry-opentelemetry/sentry-opentelemetry-core/src/main/java/io/sentry/opentelemetry/SentrySpanProcessor.java

```diff
@@ -22,14 +98,140 @@ public boolean isStartRequired() {
-    // TODO end
+  public void onEnd(final @NotNull ReadableSpan otelSpan) {
+    if (!hasSentryBeenInitialized()) {
+      return;
+    }
+
+    if (!Instrumenter.OTEL.equals(Sentry.getCurrentHub().getOptions().getInstrumenter())) {
+      return;
+    }
+
+    final @NotNull SpanContext otelSpanContext = otelSpan.getSpanContext();
+    if (!otelSpanContext.isValid()) {
+      return;
+    }
+
+    final @NotNull TraceData traceData = getTraceData(otelSpan);
+    final @Nullable ISpan sentrySpan = spans.remove(traceData.getSpanId());
+
+    if (sentrySpan == null) {
+      return;
+    }
+
+    if (isSentryRequest(otelSpan)) {
+      return;
+    }
+
+    if (sentrySpan instanceof ITransaction) {
+      ITransaction sentryTransaction = (ITransaction) sentrySpan;
+      updateTransactionWithOtelData(sentryTransaction, otelSpan);
+    } else {
+      updateSpanWithOtelData(sentrySpan, otelSpan);
+    }
+
+    final @NotNull SpanStatus sentryStatus = mapOtelStatus(otelSpan);
+    final @NotNull Date endTimestamp =
+        DateUtils.nanosToDate(otelSpan.toSpanData().getEndEpochNanos());
+    sentrySpan.finish(sentryStatus, endTimestamp);
   }
 
   @Override
   public boolean isEndRequired() {
     return true;
   }
+
+  private @NotNull TraceData getTraceData(final @NotNull ReadableSpan otelSpan) {
+    final @NotNull SpanContext otelSpanContext = otelSpan.getSpanContext();
+    final @NotNull String otelSpanId = otelSpanContext.getSpanId();
+    final @NotNull String otelParentSpanIdMaybeInvalid =
+        otelSpan.getParentSpanContext().getSpanId();
+    final @NotNull String otelTraceId = otelSpanContext.getTraceId();
+    final @Nullable String otelParentSpanId =
+        io.opentelemetry.api.trace.SpanId.isValid(otelParentSpanIdMaybeInvalid)
+            ? otelParentSpanIdMaybeInvalid
+            : null;
+
+    // TODO read parentSampled and baggage from context set by propagator
+
+    return new TraceData(otelTraceId, otelSpanId, otelParentSpanId);
+  }
+
+  private boolean isSentryRequest(final @NotNull ReadableSpan otelSpan) {
+    final @NotNull SpanKind kind = otelSpan.getKind();
+    if (!spanKindsConsideredForSentryRequests.contains(kind)) {
+      return false;
+    }
+
+    final @Nullable String httpUrl = otelSpan.getAttribute(SemanticAttributes.HTTP_URL);
+    return DsnUtil.urlContainsDsnHost(Sentry.getCurrentHub().getOptions(), httpUrl);
+  }
+
+  private void updateTransactionWithOtelData(
+      final @NotNull ITransaction sentryTransaction, final @NotNull ReadableSpan otelSpan) {
+    final @NotNull SpanDescription spanDescription =
+        spanDescriptionExtractor.extractSpanDescription(otelSpan);
+    sentryTransaction.setOperation(spanDescription.getOp());
+    sentryTransaction.setName(
+        spanDescription.getDescription(), spanDescription.getTransactionNameSource());
+
+    final @NotNull Map<String, Object> otelContext = toOtelContext(otelSpan);
+    System.out.println(otelContext);
```

Comment:
l: do we need this? If we do, can we expand the log message?

Reply:
The `System.out.println`s have all been removed in #2380 . They were just there for testing / debugging.